lib/promauth/config.go

```diff
@@ -823,3 +875,113 @@ func (tctx *tlsContext) initFromTLSConfig(baseDir string, tc *TLSConfig) error {
 	tctx.minVersion = v
 	return nil
 }
+
+// NewTLSRoundTripper returns new http.RoundTripper which automatically updates
+// RootCA in tls.Config whenever it changes.
+func (tctx *tlsContext) NewTLSRoundTripper(cfg *tls.Config, builder func(transport *http.Transport)) (http.RoundTripper, error) {
+	// TLS context is not initialized so use the provided RoundTripper without wrapper
+	if tctx == nil || tctx.getRootCAPEM == nil {
+		tr := &http.Transport{
+			TLSClientConfig: cfg,
+		}
+		builder(tr)
+		return tr, nil
+	}
+
+	var mu sync.Mutex
+	var deadline uint64
+	var rootCA []byte
+	var err error
+	getTLSDigests := func() []byte {
+		if fasttime.UnixTimestamp() > deadline {
+			mu.Lock()
+			rootCA, err = tctx.getRootCAPEM()
+			if err != nil {
+				logger.Warnf("cannot load root CA: %s", err)
+			}
+			deadline = fasttime.UnixTimestamp() + tlsCertsCacheSeconds
+			mu.Unlock()
+		}
+
+		return rootCA
+	}
+	tr := &http.Transport{
+		TLSClientConfig: cfg,
+	}
+	builder(tr)
+
+	return &TLSRoundTripper{
+		builder:        builder,
+		getRootCABytes: getTLSDigests,
+		getTLSRootCA:   newGetTLSRootCACached(tctx.getTLSRootCA),
+		config:         cfg,
+		tr:             tr,
+	}, nil
+}
+
+var _ http.RoundTripper = &TLSRoundTripper{}
+
+// TLSRoundTripper is an implementation of http.RoundTripper which automatically
+// updates RootCA in tls.Config whenever it changes.
+type TLSRoundTripper struct {
+	builder        func(*http.Transport)
+	getRootCABytes func() []byte
+	getTLSRootCA   func() (*x509.CertPool, error)
+
+	config *tls.Config
+
+	tr http.RoundTripper
+
+	m           sync.RWMutex
+	rootCABytes []byte
+}
+
+// RoundTrip implements http.RoundTripper.RoundTrip.
+// It automatically updates RootCA in tls.Config whenever it changes.
+func (t *TLSRoundTripper) RoundTrip(request *http.Request) (*http.Response, error) {
+	rootCABytes := t.getRootCABytes()
+	t.m.RLock()
```

Comment:
I would like to think about it from another perspective: http client can send hundreds/thousands RPS. In case certificate was not changed all these calls can acquire read lock without block each other.
And only once certificate has changed the write lock will be acquired. The worst case is that `RoundTripper` will be created multiple times and will replace each other.

In order to avoid this it's possible to do another check if `rootCABytes` is up-to-date after locking mutex for write. But this will mean that mutex has to be locked while creating new TLS config.

Wdyt?

Reply:
I think, it's better to rename function `getRootCABytes` for `getRootCABytesLocked` and remove mutex from it. It must be called only with obtained mutex from TLSRoundTripper. 

 Always obtain t.m.Lock and release it with different.

 It's not a hot path for application and there is no need in readwrite mutex for now.

 Later it'd be possible to rewrite with with atomic.Pointer, but I don't think that it'd be needed.

Reply:
Updated